### PR TITLE
sensord: use interrupts to improve LSM6DS3 timing accuracy

### PR DESCRIPTION
**Pull request by gast04**
Moving sensor data reading to interrupt triggered, starting with accelerometer
- enabling any data interrupt
- map interrupt to INT1, GPIO 21
- add PoC of reading the data

### CODE REVIEW

**Comment by geohot** [comment]
Are we still using the BMX? That's only present on half the devices, perhaps we should focus on the LSM.

**Comment by pd0wm** [comment]
We're just going to support interrupts for both. For devices with the BMX it's the only magnetometer on board, so we can't drop it completely.

File: selfdrive/common/gpio.h [hunk 1]
```diff
@@ -17,5 +17,9 @@
   #define GPIO_STM_BOOT0        0
 #endif
 
+enum EDGE_TYPES { rising, falling, both, none };
```
Comment (pd0wm) [inline selfdrive/common/gpio.h:20]:
```suggestion
enum Edgetypes { Rising, Falling, Both, None };
```

We're not 100% consistent in our codebase with enums lower/uppercase, but let's do this.

File: selfdrive/common/gpio.cc [hunk 2]
```diff
@@ -30,3 +30,39 @@ int gpio_set(int pin_nr, bool high) {
...
+int gpio_get_ro_value_fd(int pin_nr) {
+  char pin_dir_path[50];
+
+  int pin_dir_path_len = snprintf(pin_dir_path, sizeof(pin_dir_path),
+                          "/sys/class/gpio/gpio%d/value", pin_nr);
+  if(pin_dir_path_len <= 0) {
+    return -1;
+  }
+
+  int fd = open(pin_dir_path, O_RDONLY);
```
Comment (pd0wm) [inline selfdrive/common/gpio.cc:63]:
```suggestion
  return open(pin_dir_path, O_RDONLY);
```

open only returns -1

File: selfdrive/common/gpio.h (hunk 1 above)
Comment (gast04) [reply selfdrive/common/gpio.h:20]: sure

File: selfdrive/sensord/sensors/lsm6ds3_accel.cc [hunk 3]
```diff
@@ -27,19 +28,48 @@ int LSM6DS3_Accel::init() {
...
   return ret;
 }
 
 void LSM6DS3_Accel::get_event(cereal::SensorEventData::Builder &event) {
 
+  // INT1 shared with gyro, check STATUS_REG who triggered
+  uint8_t status_reg = 0;
+  read_register(LSM6DS3_ACCEL_I2C_REG_STAT_REG, &status_reg, sizeof(status_reg));
+  if ((status_reg & LSM6DS3_ACCEL_DRDY_XLDA) == 0) {
+    return;
```
[the remark this answers is not in the document]
Comment (gast04) [reply selfdrive/sensord/sensors/lsm6ds3_accel.cc:70]: yea thought about that, will do so

[the remark this answers is not in the document]
Comment (pd0wm) [reply selfdrive/sensord/sensors/lsm6ds3_accel.cc:70]: https://github.com/commaai/openpilot/pull/21427/files

File: selfdrive/sensord/sensors_qcom2.cc [hunk 4]
```diff
@@ -67,29 +119,43 @@ int sensor_loop() {
...
         return -1;
       }
     } else {
       if (sensor.first == &bmx055_magn || sensor.first == &mmc5603nj_magn) {
         has_magnetometer = true;
       }
-      sensors.push_back(sensor.first);
+
+      // split between interrupt read sensors and non
+      if (sensor.first->gpio_fd != -1) {
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors_qcom2.cc:138]: seems like a bit of an indirect way to determine this

File: selfdrive/sensord/sensors_qcom2.cc [hunk 5]
```diff
@@ -67,29 +119,43 @@ int sensor_loop() {
...
+      }
+      else {
+        sensors.push_back(sensor.first);
+      }
     }
   }
 
   if (!has_magnetometer) {
     LOGE("No magnetometer present");
+    delete i2c_bus_imu;
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors_qcom2.cc:149]: add a fail label instead?

File: selfdrive/sensord/sensors/bmx055_gyro.cc [hunk 6]
```diff
@@ -26,6 +27,23 @@ int BMX055_Gyro::init() {
...
+  if (gpio_set_edge(gpio_nr, Edgetypes::Rising) != 0) {
+    ret = -1;
+    goto fail;
+  }
+
+  gpio_fd = gpio_get_ro_value_fd(gpio_nr);
+  if (gpio_fd < 0) {
+    ret = -1;
+    goto fail;
+  }
```
Comment (adeebshihadeh) [inline selfdrive/sensord/sensors/bmx055_gyro.cc:45]: this block is repeated a few times, can we pull this into a helper function or something?

File: selfdrive/hardware/tici/pins.py [hunk 7]
```diff
@@ -19,3 +19,9 @@ class GPIO:
   CAM0_RSTN = 9
   CAM1_RSTN = 7
   CAM2_RSTN = 12
+
+  # GPIO mapping of sensor interrupts
```
[the remark this answers is not in the document]
Comment (gast04) [reply selfdrive/hardware/tici/pins.py]: changed

File: selfdrive/common/gpio.h [hunk 8]
```diff
@@ -17,5 +17,9 @@
   #define GPIO_STM_BOOT0        0
 #endif
 
+enum Edgetypes { Rising, Falling, Both, None };
```
[the remark this answers is not in the document]
Comment (gast04) [reply selfdrive/common/gpio.h]: changed

File: selfdrive/sensord/sensors_qcom2.cc [hunk 9]
```diff
@@ -24,6 +25,57 @@
...
+      if ((fd_list[i].revents & POLLPRI) == 0) {
+        continue;
+      }
+
+      auto orphan = orphanage.newOrphan<cereal::SensorEventData>();
+      auto event = orphan.get();
+      if (gpio_sensors[i]->get_event(event)) {
+        // only send collected events
+        collected_events.push_back(kj::mv(orphan));
+      }
```
Comment (pd0wm) [inline selfdrive/sensord/sensors_qcom2.cc:66]:
```suggestion
      char buf[64];
      lseek(fd_list[i].fd, 0, SEEK_SET);
      read(fd_list[i].fd, &buf, std::size(buf));
      
      }
```

File: selfdrive/sensord/sensors_qcom2.cc [hunk 10]
```diff
@@ -24,6 +25,57 @@
...
+    }
+
+    auto events = msg.initEvent().initSensorEvents(collected_events.size());
+    for (int i = 0; i < collected_events.size(); ++i) {
+      events.adoptWithCaveats(i, kj::mv(collected_events[i]));
+    }
+
+    pm_mutex.lock();
```
[the remark this answers is not in the document]
Comment (pd0wm) [reply selfdrive/sensord/sensors_qcom2.cc:74]: Let's do that in a separate PR together with a cereal change.

**Comment by pd0wm** [comment]
lsm_hist
bmx_hist

**Pull request closed without merging**